gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java

```diff
@@ -0,0 +1,178 @@
+ */
+
+@Slf4j
+public class LineageInfo {
+
+  public static final String LINEAGE_NAME_SPACE = "gobblin.lineage";
+  private static final String DATASET_PREFIX =  LINEAGE_NAME_SPACE + ".";
+  private static final String BRANCH_PREFIX = DATASET_PREFIX + "branch.";
+
+  @Getter
+  private String datasetUrn;
+  @Getter
+  private String jobId;
+
+  private Map<String, String> lineageMetaData;
+
+  public enum Level {
+    DATASET,
+    BRANCH,
+    All
+  }
+
+  private LineageInfo() {
+  }
+
+  private LineageInfo(String datasetUrn, String jobId, Map<String, String> lineageMetaData) {
+    Preconditions.checkArgument(datasetUrn != null);
+    Preconditions.checkArgument(jobId != null);
+    this.datasetUrn = datasetUrn;
+    this.jobId = jobId;
+    this.lineageMetaData = lineageMetaData;
+  }
+
+  /**
+   * Retrieve lineage information from a {@link State} by {@link Level}
+   * @param level {@link Level#DATASET}  only load dataset level lineage attributes
+   *              {@link Level#BRANCH}   only load branch level lineage attributes
+   *              {@link Level#All}      load all lineage attributes
+   * @return A {@link LineageInfo} object containing all lineage attributes
+   */
+  public static LineageInfo load (State state, Level level) throws LineageException {
+    Preconditions.checkArgument(state != null);
+    HashMap<String, String> metaData = new HashMap<>();
+
+    for (Map.Entry<Object, Object> entry : state.getProperties().entrySet()) {
+      if (entry.getKey() instanceof String && ((String) entry.getKey()).startsWith(LINEAGE_NAME_SPACE)) {
+
+        String lineageKey = ((String) entry.getKey());
+        String lineageValue = (String) entry.getValue();
+
+        if (lineageKey.startsWith(BRANCH_PREFIX)) {
+          if (level == Level.BRANCH || level == Level.All) {
+            metaData.put(lineageKey.substring(BRANCH_PREFIX.length()), lineageValue);
+          }
+        } else if (lineageKey.startsWith(DATASET_PREFIX)) {
+          if (level == Level.DATASET || level == Level.All) {
+            metaData.put(lineageKey.substring(DATASET_PREFIX.length()), lineageValue);
+          }
+        }
+      }
+    }
+
+    String jobId = metaData.containsKey(ConfigurationKeys.JOB_ID_KEY)? metaData.get(ConfigurationKeys.JOB_ID_KEY):"";
+    String urn = metaData.containsKey(ConfigurationKeys.DATASET_URN_KEY)? metaData.get(ConfigurationKeys.DATASET_URN_KEY):ConfigurationKeys.DEFAULT_DATASET_URN;
+
+    metaData.remove(ConfigurationKeys.JOB_ID_KEY);
+    metaData.remove(ConfigurationKeys.DATASET_URN_KEY);
+
+    LineageInfo descriptor = new LineageInfo(urn, jobId, metaData);
+
+    return descriptor;
+  }
+
+  /**
+   * Get all lineage meta data.
+   */
+  public ImmutableMap<String, String> getLineageMetaData() {
+    return ImmutableMap.copyOf(lineageMetaData);
+  }
+
+
+  public static LineageInfo load (Collection<? extends State> states, Level level) throws LineageException {
```

Comment:
Please add javadoc and specify that all states need to be part of the same branch / dataset depending on the level.